File: backend_code/upload-stages.js
const admin = require('firebase-admin')
const fs = require('fs')
const path = require('path')

// Initialize Firebase Admin
// Make sure to set GOOGLE_APPLICATION_CREDENTIALS environment variable
// or provide service account key file path
if (!admin.apps.length) {
  try {
    // Try to initialize with default credentials (from environment variable)
    admin.initializeApp({
      credential: admin.credential.applicationDefault()
    })
  } catch (error) {
    // If default credentials fail, try service account file
    const serviceAccountPath = process.env.SERVICE_ACCOUNT_PATH || './serviceAccountKey.json'
    if (fs.existsSync(serviceAccountPath)) {
      const serviceAccount = require(path.resolve(serviceAccountPath))
      admin.initializeApp({
        credential: admin.credential.cert(serviceAccount)
      })
    } else {
      console.error('Error: Firebase Admin not initialized.')
      console.error('Please set GOOGLE_APPLICATION_CREDENTIALS environment variable')
      console.error('or provide serviceAccountKey.json file in the current directory.')
      process.exit(1)
    }
  }
}

const db = admin.firestore()

async function checkStagesCollection() {
  try {
    // Verify Firestore connection by attempting to access the stages collection
    const stagesRef = db.collection('stages')
    await stagesRef.limit(1).get()
    return true
  } catch (error) {
    // If we get a permission error or connection error, collection is not accessible
    if (error.code === 7 || error.code === 16 || error.message.includes('permission') || error.message.includes('not found')) {
      return false
    }
    // For other errors, still return false to be safe
    return false
  }
}

async function uploadStages(stagesData) {
  const batch = db.batch()
  let batchCount = 0
  const BATCH_LIMIT = 500 // Firestore batch limit

  for (const stage of stagesData) {
    // Ensure stageId matches document ID format
    const docId = stage.stageId ? `stage${stage.stageId}` : stage.docId || `stage${stage.stageId}`
    
    if (!docId) {
      console.error(`Skipping stage without docId or stageId:`, stage)
      continue
    }

    const docRef = db.collection('stages').doc(docId)
    
    // Prepare document data (exclude docId from data)
    const { docId: _, ...stageData } = stage
    
    batch.set(docRef, {
      ...stageData,
      stageId: stage.stageId || parseInt(docId.replace('stage', ''))
    }, { merge: false }) // Overwrite completely
    
    batchCount++

    // Firestore batch limit is 500 operations
    if (batchCount >= BATCH_LIMIT) {
      await batch.commit()
      console.log(`Committed batch of ${batchCount} documents`)
      batchCount = 0
    }
  }

  // Commit remaining documents
  if (batchCount > 0) {
    await batch.commit()
    console.log(`Committed final batch of ${batchCount} documents`)
  }
}

async function main() {
  // Get stages data file path from command line or use default
  const stagesFile = process.argv[2] || './stages-data.json'
  
  if (!fs.existsSync(stagesFile)) {
    console.error(`Error: Stages data file not found: ${stagesFile}`)
    console.error('Usage: node upload-stages.js [path-to-stages-data.json]')
    process.exit(1)
  }

  // Read stages data
  let stagesData
  try {
    const fileContent = fs.readFileSync(stagesFile, 'utf8')
    stagesData = JSON.parse(fileContent)
  } catch (error) {
    console.error(`Error reading or parsing ${stagesFile}:`, error.message)
    process.exit(1)
  }

  // Ensure stagesData is an array
  if (!Array.isArray(stagesData)) {
    console.error('Error: Stages data must be an array')
    process.exit(1)
  }

  if (stagesData.length === 0) {
    console.error('Error: No stages data found in file')
    process.exit(1)
  }

  console.log(`Found ${stagesData.length} stage(s) to upload`)

  // Check if stages collection exists
  console.log('Checking if stages collection exists...')
  const collectionExists = await checkStagesCollection()
  
  if (!collectionExists) {
    console.error('Error: stages collection not found or not accessible')
    process.exit(1)
  }

  console.log('Stages collection found. Uploading documents...')

  try {
    await uploadStages(stagesData)
    console.log(`Successfully uploaded ${stagesData.length} stage document(s)`)
  } catch (error) {
    console.error('Error uploading stages:', error.message)
    process.exit(1)
  }
}

// Run the script
main()
  .then(() => {
    console.log('Upload complete!')
    process.exit(0)
  })
  .catch((error) => {
    console.error('Fatal error:', error)
    process.exit(1)
  })

File: backend_code/upload-stages.py
#!/usr/bin/env python3
import sys
import json
import os
from firebase_admin import credentials, firestore, initialize_app

def check_stages_collection(db):
    """Check if stages collection exists and is accessible"""
    try:
        # Verify Firestore connection by attempting to access the stages collection
        stages_ref = db.collection('stages')
        list(stages_ref.limit(1).stream())
        return True
    except Exception as e:
        # If we get a permission error or connection error, collection is not accessible
        error_str = str(e).lower()
        if 'permission' in error_str or 'not found' in error_str or 'not accessible' in error_str:
            return False
        # For other errors, still return False to be safe
        return False

def upload_stages(db, stages_data):
    """Upload stages to Firestore, overwriting existing documents"""
    batch = db.batch()
    batch_count = 0
    BATCH_LIMIT = 500  # Firestore batch limit
    
    for stage in stages_data:
        # Determine document ID
        if 'stageId' in stage:
            doc_id = f"stage{stage['stageId']}"
        elif 'docId' in stage:
            doc_id = stage['docId']
        else:
            print(f"Skipping stage without docId or stageId: {stage}")
            continue
        
        # Prepare document data (exclude docId from data)
        stage_data = {k: v for k, v in stage.items() if k != 'docId'}
        
        # Ensure stageId is set
        if 'stageId' not in stage_data:
            stage_data['stageId'] = int(doc_id.replace('stage', ''))
        
        doc_ref = db.collection('stages').document(doc_id)
        batch.set(doc_ref, stage_data)  # Overwrite completely
        batch_count += 1
        
        # Firestore batch limit is 500 operations
        if batch_count >= BATCH_LIMIT:
            batch.commit()
            print(f'Committed batch of {batch_count} documents')
            batch_count = 0
            batch = db.batch()
    
    # Commit remaining documents
    if batch_count > 0:
        batch.commit()
        print(f'Committed final batch of {batch_count} documents')

def main():
    # Get stages data file path from command line or use default
    stages_file = sys.argv[1] if len(sys.argv) > 1 else './stages-data.json'
    
    if not os.path.exists(stages_file):
        print(f'Error: Stages data file not found: {stages_file}')
        print('Usage: python upload-stages.py [path-to-stages-data.json]')
        sys.exit(1)
    
    # Read stages data
    try:
        with open(stages_file, 'r', encoding='utf-8') as f:
            stages_data = json.load(f)
    except json.JSONDecodeError as e:
        print(f'Error parsing JSON file: {e}')
        sys.exit(1)
    except Exception as e:
        print(f'Error reading file: {e}')
        sys.exit(1)
    
    # Ensure stages_data is a list
    if not isinstance(stages_data, list):
        print('Error: Stages data must be an array')
        sys.exit(1)
    
    if len(stages_data) == 0:
        print('Error: No stages data found in file')
        sys.exit(1)
    
    print(f'Found {len(stages_data)} stage(s) to upload')
    
    # Initialize Firebase Admin
    try:
        # Try to use default credentials (from GOOGLE_APPLICATION_CREDENTIALS)
        try:
            cred = credentials.ApplicationDefault()
            initialize_app(cred)
        except ValueError:
            # Already initialized, continue
            pass
        except Exception:
            # If default credentials fail, try service account file
            try:
                service_account_path = os.getenv('SERVICE_ACCOUNT_PATH', './serviceAccountKey.json')
                if os.path.exists(service_account_path):
                    cred = credentials.Certificate(service_account_path)
                    initialize_app(cred)
                else:
                    print('Error: Firebase Admin not initialized.')
                    print('Please set GOOGLE_APPLICATION_CREDENTIALS environment variable')
                    print('or provide serviceAccountKey.json file in the current directory.')
                    sys.exit(1)
            except ValueError:
                # Already initialized, continue
                pass
    except Exception as e:
        print(f'Error initializing Firebase Admin: {e}')
        sys.exit(1)
    
    db = firestore.client()
    
    # Check if stages collection exists
    print('Checking if stages collection exists...')
    if not check_stages_collection(db):
        print('Error: stages collection not found or not accessible')
        sys.exit(1)
    
    print('Stages collection found. Uploading documents...')
    
    try:
        upload_stages(db, stages_data)
        print(f'Successfully uploaded {len(stages_data)} stage document(s)')
    except Exception as e:
        print(f'Error uploading stages: {e}')
        sys.exit(1)
    
    print('Upload complete!')

if __name__ == '__main__':
    main()

